core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksStore.java

```diff
@@ -47,41 +51,81 @@
 import java.io.IOException;
 import java.io.OutputStream;
 import java.nio.file.Paths;
+import java.time.Duration;
+import java.time.Instant;
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.List;
 import java.util.Optional;
 import java.util.UUID;
+import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicReference;
-import javax.annotation.concurrent.ThreadSafe;
+import java.util.concurrent.atomic.LongAdder;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import javax.annotation.concurrent.NotThreadSafe;
 
 /**
  * Class for managing a rocksdb database. This class handles common functionality such as
  * initializing the database and performing database backup/restore.
  *
- * Thread safety is achieved by synchronizing all public methods.
+ * This class provides locking methods for the callers. And the thread safety of RocksDB
+ * relies on the caller to use the corresponding lock methods.
+ * The reasons why this class only provides thread safety utilities to the callers
+ * (instead of wrapping it under each call) are:
+ * 1. Callers like RocksInodeStore and RocksBlockMetaStore have specific read/write logic
+ *    like iteration, which cannot be abstracted and locked internally in this class.
+ * 2. With locking methods provided by this class, callers like RocksInodeStore
+ *    can actually reuse the locks to perform concurrency control on their own logic.
+ *
+ * For reading/writing on the RocksDB, use the shared lock
+ * <blockquote><pre>
+ *   try (RocksSharedLockHandle r = mRocksStore.checkAndAcquireSharedLock() {
+ *     // perform your read/write operation
+ *   }
+ * </pre></blockquote>
+ *
+ * For operations like closing/restart/restoring on the RocksDB, an exclusive lock should
+ * be acquired by calling one of:
+ * 1. {@link #lockForClosing()}
+ * 2. {@link #lockForRestart()}
  */
-@ThreadSafe
+@NotThreadSafe
```

Comment:
Before: thread safety by `synchronized` methods and `@ThreadSafe` tag, however the thread safety beyond the `synchronized` call is flawed
After: removed embedded thread safety and removed `synchronized`. The thread safety is provided by providing locking methods, and the caller should invoke those calls in the correct way.